cluster-provision/gocli/pkg/libssh/ssh.go

```diff
@@ -84,10 +85,122 @@ func (s *SSHClientImpl) Command(cmd string) error {
 			cmd = "sudo /bin/bash " + cmd
 		}
 	}
+	logrus.Infof("[node %d]: %s", s.nodeIdx, cmd)
 
 	err = session.Run(cmd)
 	if err != nil {
 		return fmt.Errorf("Failed to execute command: %v, %v", cmd, err)
 	}
 	return nil
 }
+
+func (s *SSHClientImpl) CommandWithNoStdOut(cmd string) (string, error) {
```

Comment:
It seems we have duplicates now, can we somehow have common parts?

Reply:
@xpivarc 
there's a problem with this

```golang
func (s *SSHClientImpl) executeCommand(cmd string, outBuffer, errBuffer io.Writer) (string, error) {
	session, err := s.client.NewSession()
	if err != nil {
		return "", err
	}
	defer session.Close()

	session.Stdout = outBuffer
	session.Stderr = errBuffer

	if len(cmd) > 0 {
		firstCmdChar := cmd[0]
		// indicates the command is a script or a script with params
		if string(firstCmdChar) == "/" || string(firstCmdChar) == "-" {
			cmd = "sudo /bin/bash " + cmd
		}
	}
	logrus.Infof("[node %d]: %s", s.nodeIdx, cmd)

	err = session.Run(cmd)
	if err != nil {
		err = fmt.Errorf(errBuffer.String())
		return "", fmt.Errorf("Failed to execute command: %v, %v", cmd, err)
	}
	return outBuffer.String(), nil
}
```

if you try to create a method that can take an arbitrary io writer a problem arises because how you handle both cases is dependent on what the io writer actually is. a file, or a buffer. you can combat this by type assertion. but technically an io.Discard is a writer too, should you account for it ?
it gives the feeling that the method can can an io writer but in fact will only work an os file or a buffer


```golang
func (s *SSHClientImpl) executeCommand(cmd string, outWriter, errWriter io.Writer) (string, error) {
	session, err := s.client.NewSession()
	if err != nil {
		return "", err
	}
	defer session.Close()

	session.Stdout = outWriter
	session.Stderr = errWriter

	if len(cmd) > 0 {
		firstCmdChar := cmd[0]
		// indicates the command is a script or a script with params
		if string(firstCmdChar) == "/" || string(firstCmdChar) == "-" {
			cmd = "sudo /bin/bash " + cmd
		}
	}
	logrus.Infof("[node %d]: %s", s.nodeIdx, cmd)

	if _, ok := outWriter.(*os.File); ok {
		err = session.Run(cmd)
		if err != nil {
			return "", fmt.Errorf("Failed to execute command: %v, %v", cmd, err)
		}
		return "", nil
	}

	if outbuf, ok := outWriter.(*bytes.Buffer); ok {
		errbuf, ok := errWriter.(*bytes.Buffer)
		if !ok {
			return "", fmt.Errorf("Unable to cast error writer as a buffer")
		}
		err = session.Run(cmd)
		if err != nil {
			err = fmt.Errorf(errbuf.String())
			return "", fmt.Errorf("Failed to execute command: %v, %v", cmd, err)
		}
		return outbuf.String(), nil
	}
	return "", fmt.Errorf("Invalid writers passed, must use a buffer or an os file")
}
```

let me know your thoughts

Reply:
```
diff --git a/cluster-provision/gocli/pkg/libssh/ssh.go b/cluster-provision/gocli/pkg/libssh/ssh.go
index 16f35300..d7539fa2 100644
--- a/cluster-provision/gocli/pkg/libssh/ssh.go
+++ b/cluster-provision/gocli/pkg/libssh/ssh.go
@@ -64,6 +64,10 @@ func NewSSHClient(port uint16, idx int, root bool) (*SSHClientImpl, error) {
 // SSH performs two ssh connections, one to the forwarded port by dnsmasq to the local which is the ssh port of the control plane node
 // then a hop to the designated host where the command is desired to be ran
 func (s *SSHClientImpl) Command(cmd string) error {
+	return s.executeCommand(cmd, os.Stdout, os.Stderr)
+}
+
+func (s *SSHClientImpl) executeCommand(cmd string, outWriter, errWriter io.Writer) error {
 	if s.client == nil {
 		err := s.initClient()
 		if err != nil {
@@ -77,8 +81,8 @@ func (s *SSHClientImpl) Command(cmd string) error {
 	}
 	defer session.Close()
 
-	session.Stdout = os.Stdout
-	session.Stderr = os.Stderr
+	session.Stdout = outWriter
+	session.Stderr = errWriter
 
 	if len(cmd) > 0 {
 		firstCmdChar := cmd[0]
@@ -91,43 +95,20 @@ func (s *SSHClientImpl) Command(cmd string) error {
 
 	err = session.Run(cmd)
 	if err != nil {
-		return fmt.Errorf("Failed to execute command: %v, %v", cmd, err)
+		return fmt.Errorf("failed to execute command: %s", cmd)
 	}
 	return nil
 }
 
 func (s *SSHClientImpl) CommandWithNoStdOut(cmd string) (string, error) {
-	if s.client == nil {
-		err := s.initClient()
-		if err != nil {
-			return "", err
-		}
-	}
-
-	session, err := s.client.NewSession()
-	if err != nil {
-		return "", err
-	}
-	defer session.Close()
 	var stdout, stderr bytes.Buffer
-	session.Stdout = &stdout
-	session.Stderr = &stderr
 
-	if len(cmd) > 0 {
-		firstCmdChar := cmd[0]
-		// indicates the command is a script or a script with params
-		if string(firstCmdChar) == "/" || string(firstCmdChar) == "-" {
-			cmd = "sudo /bin/bash " + cmd
-		}
-	}
-	logrus.Infof("[node %d]: %s", s.nodeIdx, cmd)
-
-	err = session.Run(cmd)
+	err := s.executeCommand(cmd, &stdout, &stderr)
 	if err != nil {
-		err = fmt.Errorf(stderr.String())
-		return "", fmt.Errorf("Failed to execute command: %v, %v", cmd, err)
+		return "", fmt.Errorf("%w, %s", err, stderr.String())
 	}
 	return stdout.String(), nil
+
 }
 
 func (s *SSHClientImpl) SCP(fileName string, contents io.Reader) error {
```

Reply:
How about this?

Reply:
hmmm, yeah that makes alot of sense.
i will add that